File: solana_toolbox_idl/src/toolbox_idl_type_primitive.rs
#[derive(Debug, Clone, PartialEq)]
pub enum ToolboxIdlTypePrimitive {
    U8,
    U16,
    U32,
    U64,
    U128,
    I8,
    I16,
    I32,
    I64,
    I128,
    F32,
    F64,
    Bool,
    Pubkey,
}

impl ToolboxIdlTypePrimitive {
    pub fn try_parse(name: &str) -> Option<ToolboxIdlTypePrimitive> {
        match name {
            "u8" => Some(ToolboxIdlTypePrimitive::U8),
            "u16" => Some(ToolboxIdlTypePrimitive::U16),
            "u32" => Some(ToolboxIdlTypePrimitive::U32),
            "u64" => Some(ToolboxIdlTypePrimitive::U64),
            "u128" => Some(ToolboxIdlTypePrimitive::U128),
            "i8" => Some(ToolboxIdlTypePrimitive::I8),
            "i16" => Some(ToolboxIdlTypePrimitive::I16),
            "i32" => Some(ToolboxIdlTypePrimitive::I32),
            "i64" => Some(ToolboxIdlTypePrimitive::I64),
            "i128" => Some(ToolboxIdlTypePrimitive::I128),
            "f32" => Some(ToolboxIdlTypePrimitive::F32),
            "f64" => Some(ToolboxIdlTypePrimitive::F64),
            "bool" => Some(ToolboxIdlTypePrimitive::Bool),
            "pubkey" => Some(ToolboxIdlTypePrimitive::Pubkey),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &str {
        match self {
            ToolboxIdlTypePrimitive::U8 => "u8",
            ToolboxIdlTypePrimitive::U16 => "u16",
            ToolboxIdlTypePrimitive::U32 => "u32",
            ToolboxIdlTypePrimitive::U64 => "u64",
            ToolboxIdlTypePrimitive::U128 => "u128",
            ToolboxIdlTypePrimitive::I8 => "i8",
            ToolboxIdlTypePrimitive::I16 => "i16",
            ToolboxIdlTypePrimitive::I32 => "i32",
            ToolboxIdlTypePrimitive::I64 => "i64",
            ToolboxIdlTypePrimitive::I128 => "i128",
            ToolboxIdlTypePrimitive::F32 => "f32",
            ToolboxIdlTypePrimitive::F64 => "f64",
            ToolboxIdlTypePrimitive::Bool => "bool",
            ToolboxIdlTypePrimitive::Pubkey => "pubkey",
        }
    }

    pub fn size(&self) -> usize {
        match self {
            ToolboxIdlTypePrimitive::U8 => 1,
            ToolboxIdlTypePrimitive::U16 => 2,
            ToolboxIdlTypePrimitive::U32 => 4,
            ToolboxIdlTypePrimitive::U64 => 8,
            ToolboxIdlTypePrimitive::U128 => 16,
            ToolboxIdlTypePrimitive::I8 => 1,
            ToolboxIdlTypePrimitive::I16 => 2,
            ToolboxIdlTypePrimitive::I32 => 4,
            ToolboxIdlTypePrimitive::I64 => 8,
            ToolboxIdlTypePrimitive::I128 => 16,
            ToolboxIdlTypePrimitive::F32 => 4,
            ToolboxIdlTypePrimitive::F64 => 8,
            ToolboxIdlTypePrimitive::Bool => 1,
            ToolboxIdlTypePrimitive::Pubkey => 32,
        }
    }

    pub fn alignment(&self) -> usize {
        match self {
            ToolboxIdlTypePrimitive::U8 => 1,
            ToolboxIdlTypePrimitive::U16 => 2,
            ToolboxIdlTypePrimitive::U32 => 4,
            ToolboxIdlTypePrimitive::U64 => 8,
            ToolboxIdlTypePrimitive::U128 => 16,
            ToolboxIdlTypePrimitive::I8 => 1,
            ToolboxIdlTypePrimitive::I16 => 2,
            ToolboxIdlTypePrimitive::I32 => 4,
            ToolboxIdlTypePrimitive::I64 => 8,
            ToolboxIdlTypePrimitive::I128 => 16,
            ToolboxIdlTypePrimitive::F32 => 4,
            ToolboxIdlTypePrimitive::F64 => 8,
            ToolboxIdlTypePrimitive::Bool => 1,
            ToolboxIdlTypePrimitive::Pubkey => 1,
        }
    }
}

File: solana_toolbox_web3/src/ToolboxIdlTypePrimitive.ts
export class ToolboxIdlTypePrimitive {
  public static readonly U8 = new ToolboxIdlTypePrimitive({
    name: 'u8',
    size: 1,
    alignment: 1,
  });
  public static readonly U16 = new ToolboxIdlTypePrimitive({
    name: 'u16',
    size: 2,
    alignment: 2,
  });
  public static readonly U32 = new ToolboxIdlTypePrimitive({
    name: 'u32',
    size: 4,
    alignment: 4,
  });
  public static readonly U64 = new ToolboxIdlTypePrimitive({
    name: 'u64',
    size: 8,
    alignment: 8,
  });
  public static readonly U128 = new ToolboxIdlTypePrimitive({
    name: 'u128',
    size: 16,
    alignment: 16,
  });
  public static readonly I8 = new ToolboxIdlTypePrimitive({
    name: 'i8',
    size: 1,
    alignment: 1,
  });
  public static readonly I16 = new ToolboxIdlTypePrimitive({
    name: 'i16',
    size: 2,
    alignment: 2,
  });
  public static readonly I32 = new ToolboxIdlTypePrimitive({
    name: 'i32',
    size: 4,
    alignment: 4,
  });
  public static readonly I64 = new ToolboxIdlTypePrimitive({
    name: 'i64',
    size: 8,
    alignment: 8,
  });
  public static readonly I128 = new ToolboxIdlTypePrimitive({
    name: 'i128',
    size: 16,
    alignment: 16,
  });
  public static readonly F32 = new ToolboxIdlTypePrimitive({
    name: 'f32',
    size: 4,
    alignment: 4,
  });
  public static readonly F64 = new ToolboxIdlTypePrimitive({
    name: 'f64',
    size: 8,
    alignment: 8,
  });
  public static readonly Bool = new ToolboxIdlTypePrimitive({
    name: 'bool',
    size: 1,
    alignment: 1,
  });
  public static readonly Pubkey = new ToolboxIdlTypePrimitive({
    name: 'pubkey',
    size: 32,
    alignment: 1,
  });

  public static readonly primitiveByName = (() => {
    let primitives = [
      ToolboxIdlTypePrimitive.U8,
      ToolboxIdlTypePrimitive.U16,
      ToolboxIdlTypePrimitive.U32,
      ToolboxIdlTypePrimitive.U64,
      ToolboxIdlTypePrimitive.U128,
      ToolboxIdlTypePrimitive.I8,
      ToolboxIdlTypePrimitive.I16,
      ToolboxIdlTypePrimitive.I32,
      ToolboxIdlTypePrimitive.I64,
      ToolboxIdlTypePrimitive.I128,
      ToolboxIdlTypePrimitive.F32,
      ToolboxIdlTypePrimitive.F64,
      ToolboxIdlTypePrimitive.Bool,
      ToolboxIdlTypePrimitive.Pubkey,
    ];
    let primitivesByName = new Map<string, ToolboxIdlTypePrimitive>();
    for (let primitive of primitives) {
      primitivesByName.set(primitive.name, primitive);
    }
    return primitivesByName;
  })();

  public name: string;
  public size: number;
  public alignment: number;

  private constructor(value: {
    name: string;
    alignment: number;
    size: number;
  }) {
    this.name = value.name;
    this.size = value.size;
    this.alignment = value.alignment;
  }

  public traverse<P1, P2, T>(
    visitor: {
      u8: (param1: P1, param2: P2) => T;
      u16: (param1: P1, param2: P2) => T;
      u32: (param1: P1, param2: P2) => T;
      u64: (param1: P1, param2: P2) => T;
      u128: (param1: P1, param2: P2) => T;
      i8: (param1: P1, param2: P2) => T;
      i16: (param1: P1, param2: P2) => T;
      i32: (param1: P1, param2: P2) => T;
      i64: (param1: P1, param2: P2) => T;
      i128: (param1: P1, param2: P2) => T;
      f32: (param1: P1, param2: P2) => T;
      f64: (param1: P1, param2: P2) => T;
      bool: (param1: P1, param2: P2) => T;
      pubkey: (param1: P1, param2: P2) => T;
    },
    param1: P1,
    param2: P2,
  ): T {
    return visitor[this.name as keyof typeof visitor](param1, param2);
  }
}
